services/crawler/crawler.go

```diff
@@ -0,0 +1,102 @@
+package crawler
+
+import (
+	"strings"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/beego/beego/v2/core/logs"
+)
+
+var doc *goquery.Document
+var htmlResponse string
+var selectorMapping = map[string]string{
+	"topAdWordAdvertisersCount":   "#tads .uEierd",
+	"topAdWordAdvertisersUrls":    "#tads .Krnil",
+	"totalAdWordAdvertisersCount": ".Krnil",
+	"resultsCount":                "#rso .yuRUbf",
+	"resultsUrls":                 "#rso .yuRUbf > a",
+	"totalLinksCount":             "a",
+}
+
+type CrawlData struct {
+	TopAdWordAdvertisersCount   int
+	TopAdWordAdvertisersUrls    []string
+	TotalAdWordAdvertisersCount int
+	ResultsCount                int
+	ResultsUrls                 []string
+	TotalLinksCount             int
+	Html                        string
+}
+
+// Call this function with your search key and base url, it will return the necessary crawled data
+// Ex: Crawl("Buy laptop", "baseUrl")
+func Crawl(searchString string, baseUrl string) (data *CrawlData, err error) {
```

Comment:
I do not see the value in having `baseUrl` as a parameter i.e. I do see the advantage to have `GoogleSearchBaseUrl` defined as a global constant. All the selectors in `selectorMapping` as tied to Google. Therefore, it's not like it would work other search engine e.g. https://www.bing.com/search 🤔 So having it as a package constant would make it easier and fit for the current requirements.

Reply:
> I do not see the value in having baseUrl as a parameter

Me too 😄 , but I am passing the dependency so I can test various scenarios(invalid url/url parsing fails. Without passing the defenses a bit harder to test differences scenarios. Then I found this answer while the author suggested to pass the defenses as parameter so we can test them easily https://stackoverflow.com/a/19168875.

What is your thought?

Reply:
The example shows the best practice of dependency injection (see the usage of an interface). However, in your code, I don't see it as dependency injection as it's a static value (not an interface). Also, I don't see the cases of when this is mocked 🙃 It's only used in the test "given the base URL is invalid" but that test is not very valuable. If the search URL is static and not a parameter then this case would not exist. So I still recommend to remove this parameter and using a package specific constant.

Reply:
I removed the dependency injection here https://github.com/junan/google-scraper/pull/29/commits/6f5291a8cd663ce6b0500d6bfc275826273bdf85